org.eclipse.text.quicksearch/src/org/eclipse/text/quicksearch/internal/core/LineItem.java

```diff
@@ -26,19 +30,50 @@ public class LineItem {
 	int lineOffset;
 
 	public LineItem(IFile f, String line, int lineNumber, int lineOffset) {
-		this.f = f;
+		this.f = getInnermostProjectFile(f);
 		this.line = line;
 		this.lineNumber = lineNumber;
 		this.lineOffset = lineOffset;
 	}
 
 	public LineItem(FileMatch match) {
-		this.f = match.getFile();
+		this.f = getInnermostProjectFile(match.getFile());
 		this.line = match.getLineElement().getContents();
 		this.lineNumber = match.getLineElement().getLine();
 		this.lineOffset = match.getLineElement().getOffset();
 	}
 
+	private IFile getInnermostProjectFile(IFile file) {
+		// In the case where projects are nested, a text search will find a result more than once.
+		// We want to refine the outer project(s) matches to avoid duplicates but also because an outer match
+		// when opened will use an editor chosen by the outer project which may be incorrect (e.g. outer
+		// project is generic and inner project is Java).  So, we convert the file to be the IResource found in
+		// the innermost project.  This fixes: Bug 559340  https://bugs.eclipse.org/bugs/show_bug.cgi?id=559340
+		IProject[] projects = ResourcesPlugin.getWorkspace().getRoot().getProjects();
+		IPath filePath = file.getLocation();
+		IProject fileProject = file.getProject();
+		IPath currentPath = fileProject.getLocation();
+		IProject innermostProject = file.getProject();
+		for (IProject project : projects) {
+			if (project == fileProject) {
+				continue;
+			}
+			IPath projectPath = project.getLocation();
+			if (projectPath.isPrefixOf(filePath) && currentPath.isPrefixOf(projectPath)) {
+				currentPath = projectPath;
+				innermostProject = project;
+			}
+		}
+		if (innermostProject != fileProject) {
+			IPath newFilePath = filePath.removeFirstSegments(currentPath.segmentCount());
+			IFile newFile = innermostProject.getFile(newFilePath);
+			if (newFile.exists()) {
+				file = newFile;
+			}
+		}
+		return file;
```

Comment:
```suggestion  
		URI locationUri = resource.getLocationURI();
		return locationUri == null ? resource : //
			Arrays.stream(resource.getWorkspace().getRoot().findFilesForLocationURI(locationUri)) //
				.min(Comparator.comparingInt(aFile -> aFile.getFullPath().segments().length))  // shortest workspace (project reletive) full path means most nested project
				.orElse(resource);
```

Reply:
Thanks Mickael.  I used your suggestion but had to modify the resource name to be file as passed in to the method.